docs/proposals/generic-schema-batching-proposal.md

```diff
@@ -0,0 +1,987 @@
+- `isBatchable(schemaId): boolean` - Quick batchability check
+- `invalidateSchema(schemaId): Promise<void>` - Admin cache clear
+
+#### 2. Generic Content DTO
+
+**Purpose**: Accept content for any schema without type-specific constraints.
+
+**Location**: `libs/types/src/dtos/content-publishing/generic-content.dto.ts`
+
+**Structure**:
+```typescript
+GenericContentDto {
+  schemaId: number;           // 1-65536, validated
+  content: Record<string, any>; // Pre-formatted per schema
+  referenceId?: string;       // Optional tracking ID
+}
+
+GenericContentResponseDto {
+  referenceId: string;        // For tracking
+  schemaId: number;           // Confirms schema
+  batched: boolean;           // Whether batched or direct
+}
+```
+
+**Key Point**: Content must be **pre-formatted** to match the target schema's structure. The Gateway does no transformation in Solution 1.
+
+#### 3. Generic API Endpoint
+
+**Location**: `apps/content-publishing-api/src/controllers/v1/content.controller.v1.ts`
+
+**New Endpoint**: `POST /v1/content/:msaId/generic`
+
+**Flow**:
+```
+1. Receive request with schemaId + content
+2. Validate schema exists via SchemaRegistry
+3. Check if schema is batchable
+4. If batchable → queue for batching (with schema-specific queue name)
+5. If not batchable → queue for direct on-chain publishing
+6. Return 202 Accepted with reference ID
+```
+
+**Example Request (AT Protocol)**:
+```json
+POST /v1/content/12345/generic
+{
+  "schemaId": 100,
+  "content": {
+    "text": "Hello from AT Protocol!",
+    "createdAt": "2025-01-15T10:00:00Z",
+    "langs": ["en"]
+  }
+}
+```
+
+#### 4. Batching Service Modification
+
+**Change Required**: Extract `schemaId` from announcement instead of hard-coded lookup.
+
+**Current Behavior** (Hard-coded):
+```typescript
+// Queue name → schema lookup
+schemaId = await getSchemaIdByName('dsnp', queueNameMap[queueName])
+```
+
+**New Behavior** (Dynamic):
+```typescript
+// Schema ID embedded in announcement
+schemaId = announcement.schemaId || fallbackToLegacyLookup()
+```
+
+**Queue Naming**: Generic schemas use pattern `batch:schema:{schemaId}` to avoid conflicts.
+
+### Data Flow Diagram
+
+```
+┌─────────────┐
+│ User Request│
+│ schemaId:100│
+└──────┬──────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ SchemaRegistry.getSchema │
+│ • Check memory cache     │
+│ • Check Redis cache      │
+│ • Fetch from blockchain  │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ SchemaDefinition         │
+│ • batchable: true        │
+│ • fullName: atproto...   │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Route Decision           │
+│ if batchable:            │
+│   → Batching Queue       │
+│ else:                    │
+│   → Direct Publish       │
+└──────┬───────────────────┘
+       │
+       ▼
+┌──────────────────────────┐
+│ Queue: batch:schema:100  │
+│ • Group by schema ID     │
```

Comment:
Do you imagine a timeout for publishing a batch if there are no other messages for a batchable schema?